### DISCUSSION TITLE
Which version of the NUnit-Console should be paired with which version of the engine?

### DISCUSSION THREAD

**Discussion by James654987** [nunit/nunit-console#1240] [Q&A] [answered]
I hope this is just a simple question and the right place to ask this.

In my project, I have NUnit 3.12.0 and NUnit3TestAdapter 4.1.0 installed. We use TeamCity as our build server and I (think I) need to install the NUnit-Console runner. I'm wondering which version I should choose to correspond correctly with the version I have installed in my project.

According to Charlie Poole's [blog](http://charliepoole.org/technical/nunit-engine-version-conflicts-in-visual-studio.html), 

> the engine version is always the same as the console package version.

which is quite clear, and tells me I should choose NUnit-Console runner version 3.12. However I'm wondering how much has changed since this blog post since the NUnit-Console version is up to 3.15.2 at the time of this writing, whereas the NUnit framework package is only up to 3.13.3. 

Does the engine version included in the console runner no longer correspond to the package version number? If so, when did they diverge? Does the console runner 3.12 have engine version 3.12 or is it different now?

**Comment by CharliePoole** [comment, marked as answer]
As you may have noticed, I updated that blog post as of March and and haven't checked the engine versions used in the adapter since then - I no longer work on the adapter, so I only check when a question comes up and I have time to do it.

However, for the console runner side, there is no change. The engine version and console version are always the same and match the package version. Separating them  would be a big change and it's not clear that any benefit would come from it unless actual engine development were to be separated from the console.

**Reply by James654987** [reply]
Thanks for the quick reply! Apologies if this is a silly question, but is the framework version different from the engine version?

**Comment by CharliePoole** [comment]
Not silly at all! The framework is versioned separately from the engine and console. The engine and any runner using it is able to work with both past and future versions of the framework because the API is standardized. This should not change until the framework does a major release.

**Reply by James654987** [reply]
Therein lies my confusion. I thought the framework and engine versions were the same, so I didn't understand how the console could use engine version 3.15 when the framework was only up to 3.13. It makes sense now. Thanks for the help!
